### PR TITLE
Add preliminary support for loom

### LINKED ISSUE #486: Add loom support to crossbeam-epoch
It'd be really cool if `crossbeam-epoch` had support for [`loom`](https://github.com/tokio-rs/loom). That would allow us to write more reliable tests where every execution interleaving and atomic visibility case actually gets exercised, without resorting to some of the "spin up a bunch of threads" tests we have today. It would also allow downstream crates to write similar tests on top of `crossbeam-epoch`, such as in `flurry` where we need loom to exercise some tricky [contention cases](https://github.com/jonhoo/flurry/pull/72#issuecomment-610618109) that are hard to hit reliably otherwise.

The loom README has a decent amount of info on how to go about adding loom support, but it _basically_ boils down to using types from `loom` instead of from `std` (or `core`) when `cfg(loom)`. Then, for tests, you just surround the test with `loom::model(|| {})`, and loom will automatically run the closure under different concurrent execution schedules.

/cc @jeehoonkang @carllerche @domenicquirl

### PR DESCRIPTION
**Pull request by jonhoo**
This patch only adds support to parts of `utils` and to `epoch`. Some
parts of `utils` had to be left out, since they rely on
`AtomicUsize::new` being `const` (which it is not in `loom`). Other
parts had to be left out due to the lack of `thread::Thread` in `loom`.
All the parts needed for `epoch` were successfully moved to loom.

The one loom test currently passes! I've added a loom pass on
crossbeam-epoch to CI.

Also, note that the uses of `UnsafeCell` in `utils` have not been moved
to `loom::cell::UnsafeCell`, as loom's `UnsafeCell` does not support `T:
?Sized`, which `AtomicCell` depends on.

Fixes #486.

### CODE REVIEW

File: crossbeam-epoch/src/lib.rs
Comment (cynecx) [inline crossbeam-epoch/src/lib.rs:86]: So this would weaken `atomic::fence(Ordering::SeqCst)`s which crossbeam-epoch uses in some places. I was wondering what kind of effects that would have, however I am not sure how hard it is to trigger it with loom?

**Comment by cynecx** [comment]
I think it would be nice if we could run some variant of the examples (`treiber-stack`) through loom too?

File: crossbeam-epoch/src/lib.rs
Comment (jonhoo) [reply crossbeam-epoch/src/lib.rs:86]: Yes, it would indeed, but I found no other way to get around it at the moment. I filed https://github.com/tokio-rs/loom/issues/117 for things like this that are currently short-comings of this PR.

**Comment by jonhoo** [comment]
Should be quite possible! The easiest way would probably be to factor out the core of `treiber-stack` into some kind of function that we could easily call from a test. Either than, or we copy-paste the treiber stack into a test and wrap it in `loom::model`.

**Comment by jonhoo** [comment]
I added one! And interestingly enough, it _fails_ with
> Causality violation: Concurrent load and mut accesses.

**Comment by jonhoo** [comment]
@carllerche When I get the error above, how do I then find out _where_ the violation happened?

**Comment by jonhoo** [comment]
I chatted to Carl, and it looks like the cause for this is probably a bug in loom: https://github.com/tokio-rs/loom/issues/124. Let's see when it is fixed whether it then passes!

**Comment by jonhoo** [comment]
Oooh, with https://github.com/tokio-rs/loom/pull/125, loom actually runs for the treiber stack example! Letting it go through the iterations now to see if it finds any problems. None so far :D

**Comment by jonhoo** [comment]
And with https://github.com/tokio-rs/loom/pull/128, it runs to completion without errors! :open_mouth:

**Comment by jonhoo** [comment]
CI doesn't like the patch in 08b37abff58ff2b92525585e8a91c5b2e2eee0c5 :sweat_smile: It will go away once loom gets another release.

**Comment by jonhoo** [comment]
Hmm, how can I see what the errors were @taiki-e? All of them just say "This check failed" as far as I can tell?

**Comment by taiki-e** [comment]
@jonhoo  I restarted CI -- looks like it fixed the issue: https://github.com/crossbeam-rs/crossbeam/pull/487/checks

**Comment by jonhoo** [comment]
The minimal version bump to 1.30 is a little awkward, but is needed to allow us to use paths for macro imports. While digging into it, I also noticed that _technically_ 1.28 isn't the msrv at all for these crates. For example, the `alloc` crate didn't stabilize [until 1.36](https://caniuse.rs/features/alloc), yet we use it if someone does a `no_std` build with `--feature alloc`..

**Comment by jonhoo** [comment]
@jeehoonkang I believe this is now ready for review :)

File: crossbeam-epoch/src/lib.rs
Comment (taiki-e) [inline crossbeam-epoch/src/lib.rs:113]: Why this is a new type, not just reexport?

File: crossbeam-utils/Cargo.toml
Comment (taiki-e) [inline crossbeam-utils/Cargo.toml:28]: I'm surprised that this works. (I felt it was a bug, but actually, [only some things like `cfg(feature)` don't seem to work.](https://github.com/rust-lang/cargo/pull/7660/files#diff-2ab1ff723e1d553d9d5382510b10cfe6R75-R97))

**Comment by taiki-e** [comment]
> the `alloc` crate didn't stabilize [until 1.36](https://caniuse.rs/features/alloc), yet we use it if someone does a `no_std` build with `--feature alloc`..

I think this is fine given that `alloc` feature is not enabled by default and its name is the same as the name of the crate on which it depends (other features should also be fine). but I agree with it is preferable to mention features that have a different msrv from the default feature.

File: crossbeam-epoch/src/lib.rs
Comment (jonhoo) [reply crossbeam-epoch/src/lib.rs:113]: I'm not sure I follow — I need to add new methods to it (`with` and `with_mut`), which requires a newtype.

File: crossbeam-utils/Cargo.toml
Comment (jonhoo) [reply crossbeam-utils/Cargo.toml:28]: Yeah, I was surprised too! I think the thinking here is that you can only use `cfg()` in this spot for `cfg` that are available from the very early stages of compilation (like `RUSTFLAGS`). Or something. Things that are only available later (like after figuring out what features the crate will be compiled by) can't be used.

File: crossbeam-epoch/src/lib.rs
Comment (taiki-e) [reply crossbeam-epoch/src/lib.rs:113]: Oops, sorry! I somehow missed that UnsafeCell does not have those methods.

Comment (jonhoo) [reply crossbeam-epoch/src/lib.rs:113]: Yeah, loom's `UnsafeCell` has a slightly different API basically because it wants to track how long you use the mutable reference for. And so we need to shim that in the non-loom case.

**Comment by cynecx** [comment]
Ping.

**Comment by jonhoo** [comment]
@cynecx I believe this is ready to merge once reviewed. I've pushed a rebased version now to make sure there aren't any conflicts.

**Comment by jonhoo** [comment]
The two failures appear unrelated to this PR. I think CI is simply broken on `master`. Once that's fixed I'd be happy to rebase again.

**Comment by jeehoonkang** [comment]
Crossbeam has evolved quite a lot recently, so I wonder if the unrelated failures are gone.

**Comment by jonhoo** [comment]
@jeehoonkang rebased!

**Comment by jonhoo** [comment]
Hey, it works again! :tada:

**Review by taiki-e — CHANGES_REQUESTED** [review]
Sorry for the delay. Could you do the following?
- Resolve conflicts.
- Rename `cfg(loom)` to something like `cfg(loom_crossbeam)` to avoid conflict with other crates that using `cfg(loom)`. (I don't know what the `loom`'s recommendation for this is. If there is a recommendation, it is better to use it.)
- Bump `loom` to 0.4.
- Document that `cfg(loom_)` is not stable API (probably on the line above `cfg(loom).dependencies`, or on the `concurrency` module).
   As for the content of the docs, I think it's okay to with something similar to the [`nightly`](https://github.com/crossbeam-rs/crossbeam/blob/47110fd42d9b5d67548071e32714f39f59af642b/crossbeam-epoch/Cargo.toml#L30-L34) or [`sanitize`](https://github.com/crossbeam-rs/crossbeam/pull/591/files#diff-01093eefa5ed07f78fe117f0a03c931349d2ab37aae2a0483e470f25c1bfe5e5R27-R29) feature.

Ideally, I'd like @jeehoonkang to review the change to epoch, but this patch doesn't change the behavior of epoch when `cfg(loom)` is disabled, so I'll merge this once my review is addressed.

File: crossbeam-epoch/src/atomic.rs
Comment (jeehoonkang) [inline crossbeam-epoch/src/atomic.rs:511]: I'd like loom to support into_inner.  I believe matching the APIs with std is one of the design goals of loom, right?

File: crossbeam-epoch/src/atomic.rs
Comment (jeehoonkang) [inline crossbeam-epoch/src/atomic.rs:154]: The reason is loom doesn't support const fn, right?  Does it now (at version 0.4)?

File: crossbeam-epoch/src/collector.rs
Comment (jeehoonkang) [inline crossbeam-epoch/src/collector.rs:112]: Tests are excluded from loom.  Is it intended?

File: crossbeam-epoch/src/internal.rs
Comment (jeehoonkang) [inline crossbeam-epoch/src/internal.rs:38]: UnsafeCell is not technically a concurrent thing.  Is it logically correct to have it under `crate::concurrency`?

File: crossbeam-epoch/src/lib.rs
Comment (jeehoonkang) [inline crossbeam-epoch/src/lib.rs:114]: It seems loom's unsafecell has these methods but std's doesn't.  I'd like to ask if you could write down a few comments on it.

File: crossbeam-epoch/tests/loom.rs
Comment (jeehoonkang) [inline crossbeam-epoch/tests/loom.rs:49]: Maybe we can refactor them to have only one copy of stack-

**Review by jeehoonkang — CHANGES_REQUESTED** [review]
Thank you for this PR!  I left a few comments.

File: crossbeam-epoch/src/atomic.rs
Comment (jonhoo) [reply crossbeam-epoch/src/atomic.rs:154]: Well, sort of. Loom's `AtomicUsize::new` is simply not `const` because it has to interact with the loom "runtime". Unlike `AtomicUsize::new` from `std`, which is really just a constant literal, loom's version must register the `AtomicUsize` with the model so that it's execution pattern can be tracked.

File: crossbeam-epoch/src/collector.rs
Comment (jonhoo) [reply crossbeam-epoch/src/collector.rs:112]: Yes indeed. When we're running under loom, we _only_ want to run the loom tests that have the loom "runtime" (in the form of `loom::model`).

File: crossbeam-epoch/src/internal.rs
Comment (jonhoo) [reply crossbeam-epoch/src/internal.rs:38]: I agree, but `concurrency` is the module that holds all the stuff that differs between the loom and non-loom contexts, and it felt weird to break out a new, dedicated top-level module just for `UnsafeCell`. We could rename `concurrency` to something else, but I can't immediately think of something better? And to give a (weak) defense for the current scheme, `UnsafeCell` does _usually_ come up in the context of concurrency where you want to unsafely assert that you can access a given type that is technically shared.

File: crossbeam-epoch/src/atomic.rs
Comment (jonhoo) [reply crossbeam-epoch/src/atomic.rs:511]: Unfortunately there are still some holes, and `into_inner` is one of them: https://github.com/tokio-rs/loom/issues/117

File: crossbeam-epoch/tests/loom.rs
Comment (jonhoo) [reply crossbeam-epoch/tests/loom.rs:49]: Hmm, I don't know of a way to grab code from an example into another file. We could move the code from the example file into the library, but then it'd no longer be an example. Maybe we could do some hacking with `include!`, but I suspect that would be far uglier. If you have ideas I'm all ears!

**Comment by jonhoo** [comment]
@taiki-e Done! Loom tests still pass (in ~11m):
```
Completed in 9690926 iterations
test treiber_stack ... ok
```

**Comment by jonhoo** [comment]
@jeehoonkang I think I addressed your comments, but will leave it to you as to whether you're satisfied with my answers :)

File: crossbeam-epoch/Cargo.toml
Comment (taiki-e) [inline crossbeam-epoch/Cargo.toml:45]: > This is not part of the public API

I thought you were trying to make this an unstable public API used by the downstream crate (https://github.com/crossbeam-rs/crossbeam/issues/486#issuecomment-610977956).
If so,  "An internal configuration" and "not part of the public API"  are not a proper description, and perhaps something like "An unstable configuration" and "not part of the stable public API" preferred.

File: crossbeam-epoch/src/atomic.rs
Comment (taiki-e) [inline crossbeam-epoch/src/atomic.rs:351]:
```suggestion
    #[const_fn(all(feature = "nightly", not(loom_crossbeam)))]
```

Comment (taiki-e) [reply crossbeam-epoch/src/atomic.rs:351]: >     #[const_fn(all(feature = "nightly", not(loom_crossbeam)))]

Oh, I haven't implemented https://github.com/taiki-e/const_fn/issues/28 yet. Updated suggestion.

File: crossbeam-epoch/tests/loom.rs
Comment (jeehoonkang) [reply crossbeam-epoch/tests/loom.rs:49]: How about just move the treiber's stack to this file?  I'm fine with removing the examples altogether.  @taiki-e ?

**Review by jeehoonkang — CHANGES_REQUESTED** [review]

File: crossbeam-epoch/tests/loom.rs
Comment (taiki-e) [reply crossbeam-epoch/tests/loom.rs:49]: I'm also fine with removing the example.

File: crossbeam-epoch/src/internal.rs
Comment (jonhoo) [reply crossbeam-epoch/src/internal.rs:38]: Hmm, yeah, it's really anything that ends up being a synchronization point. We could use `synchronization`?

**Comment by jonhoo** [comment]
I've addressed all comments beyond the naming :)

File: crossbeam-epoch/src/atomic.rs
Comment (taiki-e) [inline crossbeam-epoch/src/atomic.rs:643]:
```suggestion
            // FIXME: loom does not yet support into_inner, so we use unsync_load for now,
```

File: crossbeam-utils/Cargo.toml
Comment (taiki-e) [inline crossbeam-utils/Cargo.toml:38]:
```suggestion
# Enable the use of loom for concurrency testing.
#
# This configuration option is outside of the normal semver guarantees: minor
# versions of crossbeam may make breaking changes to it at any time.
```

File: ci/crossbeam-epoch-loom.sh
Comment (jeehoonkang) [inline ci/crossbeam-epoch-loom.sh:8]: I'd like to ask if you could write here why LOOM_MAX_PREEMPTIONS=2 is good enough.  (Or =3 is too costly.)

**Review by jeehoonkang — COMMENTED** [review]
Again, I sincerely thank you for this effort.  Now I'm mostly fine with this PR.

- I left a few minor comments that can be easily addressed.

- Regarding the name of `crate::concurrency`: I think the module is not fundamentally about concurrency but about the core platform functionalities crossbeam depends on.  So I suggest renaming it to `crate::core`, `crate::basis`, or something that represents that idea.

**Comment by jonhoo** [comment]
I went with `primitive`, though I think the word `overloaded` could also work. I went with `primitive` over `overloaded` since the types are only _sometimes_ overloaded, but including overloaded in the name might make a reader think it is _always_ overloaded.

**Review by taiki-e — APPROVED** [review]
Looks great to me! Thanks so much for working on this!

**Comment by taiki-e** [comment]
Hmm, seems there are conflicts:
crossbeam-487
@jonhoo Could you rebase?

**Comment by jonhoo** [comment]
Err, rebasing this will probably be pretty painful. Do you mean squash? I merged with master just now, and that went off without a hitch, so it should just be a matter of using GitHub's "squash and merge" button I think.

**Comment by taiki-e** [comment]
Oh, I didn't notice about "squash and merge" worked. I'll use that. Thanks for pointing that out.

**Comment by taiki-e** [comment]
@jeehoonkang: Do you want another chance at reviewing this before I merge it?

**Comment by taiki-e** [comment]
> * Rename `cfg(loom)` to something like `cfg(loom_crossbeam)` to avoid conflict with other crates that using `cfg(loom)`. (I don't know what the `loom`'s recommendation for this is. If there is a recommendation, it is better to use it.)

Oh, I noticed that loom has a PR that tries to recommend the use of cfg like `${project_name}_loom`, not `loom_${project_name}`. 
@jonhoo That PR hasn't been merged yet, but should we probably follow it?

**Comment by jonhoo** [comment]
I don't feel strongly whether the name goes first or last, but I agree that crossbeam should probably follow whatever convention ends up being documented.

**Comment by taiki-e** [comment]
I've filed #658 to rename cfg(loom_crossbeam) to cfg(crossbeam_loom). Also, I've approved tokio-rs/loom#159.

**Comment by matklad** [comment]
Is `[target.'cfg(crossbeam_loom)'.dependencies]` in Cargo.toml the right pattern here? With this, loom (and its deps) go into rev-dep Cargo.locks unconditionally. If this were a Cargo feature (`--private--loom`), then it would be invisible to consumers. 

Context: was wondering about 

```
  Downloaded generator v0.6.24
  Downloaded loom v0.4.0
  Downloaded 2 crates (81.8 KB) in 1.08s
```

output when building rust-analyzer.

**Comment by taiki-e** [comment]
@matklad 
> Is `[target.'cfg(crossbeam_loom)'.dependencies]` in Cargo.toml the right pattern here?

Yes. see Alex's [this comment](https://github.com/rust-lang/cargo/pull/7660#issuecomment-561733620).

> With this, loom (and its deps) go into rev-dep Cargo.locks unconditionally. 

As far as I know, it will be downloaded but not actually compiled. See also https://github.com/tokio-rs/bytes/issues/400 and https://github.com/tokio-rs/bytes/issues/411.
That said, I agree that the cargo messages and crates.io UI are confusing...

> If this were a Cargo feature (`--private--loom`), then it would be invisible to consumers.

Unfortunately, as far as I know, (at least in the v1 resolver) the cargo feature doesn't provide enough functionality to support loom. See also https://github.com/crossbeam-rs/crossbeam/pull/638

**Comment by matklad** [comment]
>As far as I know, it will be downloaded but not actually compiled.

compilation is not the only cost: I read Cargo.lock from time to time, and I read .lock diffs regularly. I care a lot about keeping those as small as possible. 

>Unfortunately, as far as I know, (at least in the v1 resolver) the cargo feature doesn't provide enough functionality to support loom.

Can `loom` then be gated by *both* feature and cfg? The following seems to work:

```
[target.'cfg(loom)'.dependencies]
loom = {version="*", optional=true}
```

**Comment by taiki-e** [comment]
> Can `loom` then be gated by _both_ feature and cfg? The following seems to work:
> 
> ```
> [target.'cfg(loom)'.dependencies]
> loom = {version="*", optional=true}
> ```

Oh, I didn't know it works. I would like to switch to it if it works.

**Comment by taiki-e** [comment]
will be fixed by #666

**Pull request merged**
